trie/stacktrie.go

```diff
@@ -0,0 +1,361 @@
+		}
+
+		n.val = n.hash()
+		n.nodeType = hashedNode
+		n.key = nil
+
+		// Create a leaf for the inserted part
+		o := newLeaf(st.keyOffset+diffidx+1, key, value, st.db)
+
+		// Insert both child leaves where they belong:
+		origIdx := st.key[diffidx]
+		newIdx := key[diffidx+st.keyOffset]
+		p.children[origIdx] = n
+		p.children[newIdx] = o
+		st.key = st.key[:diffidx]
+
+	case leafNode: /* Leaf */
+		// Compare both key chunks and see where they differ
+		diffidx := st.getDiffIndex(key)
+
+		// Overwriting a key isn't supported, which means that
+		// the current leaf is expected to be split into 1) an
+		// optional extension for the common prefix of these 2
+		// keys, 2) a fullnode selecting the path on which the
+		// keys differ, and 3) one leaf for the differentiated
+		// component of each key.
+		if diffidx >= len(st.key) {
+			panic("Trying to insert into existing key")
+		}
+
+		// Check if the split occurs at the first nibble of the
+		// chunk. In that case, no prefix extnode is necessary.
+		// Otherwise, create that
+		var p *StackTrie
+		if diffidx == 0 {
+			// Convert current leaf into a branch
+			st.nodeType = branchNode
+			p = st
+			st.children[0] = nil
+		} else {
+			// Convert current node into an ext,
+			// and insert a child branch node.
+			st.nodeType = extNode
+			st.children[0] = NewStackTrie(st.db)
+			st.children[0].nodeType = branchNode
+			st.children[0].keyOffset = st.keyOffset + diffidx
+			p = st.children[0]
+		}
+
+		// Create the two child leaves: the one containing the
+		// original value and the one containing the new value
+		// The child leave will be hashed directly in order to
+		// free up some memory.
+		origIdx := st.key[diffidx]
+		p.children[origIdx] = newLeaf(diffidx+1, st.key, st.val, st.db)
+		p.children[origIdx].convertToHash()
+
+		newIdx := key[diffidx+st.keyOffset]
+		p.children[newIdx] = newLeaf(p.keyOffset+1, key, value, st.db)
+
+		// Finally, cut off the key part that has been passed
+		// over to the children.
+		st.key = st.key[:diffidx]
+		st.val = nil
+	case emptyNode: /* Empty */
+		st.nodeType = leafNode
+		st.key = key[st.keyOffset:]
+		st.val = value
+	case hashedNode:
+		panic("trying to insert into hash")
+	default:
+		panic("invalid type")
+	}
+}
+
+func (st *StackTrie) hash() []byte {
+	/* Shortcut if node is already hashed */
+	if st.nodeType == hashedNode {
+		return st.val
+	}
+	// The 'hasher' is taken from a pool, but we don't actually
+	// claim an instance until all children are done with their hashing,
+	// and we actually need one
+	var h *hasher
+	var n node
+
+	switch st.nodeType {
+	case branchNode:
+		fn := &fullNode{}
+		for i, v := range st.children {
+			if v != nil {
+				// Write a 32 byte list to the sponge
+				childhash := v.hash()
+				if len(childhash) < 32 {
+					fn.Children[i] = rawNode(childhash)
+
+				} else {
+					fn.Children[i] = hashNode(childhash)
+				}
+				st.children[i] = nil // Reclaim mem from subtree
+			}
+		}
```

Comment:
This could be rewritten as 
```golang
		var nodes [17]node

		for i, child := range st.children {
			if child != nil {
				childhash := child.hash()
				if len(childhash) < 32 {
					nodes[i] = rawNode(childhash)
				}else{
					nodes[i]= hashNode(childhash)
				}
			} else {
				nodes[i] = nilValueNode
			}
		}
		nodes[16] = nilValueNode
...
		if err := rlp.Encode(&h.tmp, nodes); err != nil {
			panic(err)
		}

```

As it is now, if you look  at `fullNode.EncodeRLP`, it basically does this whole dance again, iterating over the children to produce a `[]node` replacing `nil` will `nilValueNodes`. So that can be skipped. 

Doesn't impact the benchmark a whole lot, but might do a bit of a difference on a larger trie?

Reply:
... so essentially
```golang
	case branchNode:
		var nodes [17]node
		for i, child := range st.children {
			if child == nil{
				nodes[i] = nilValueNode
				continue
			}
			if childhash := child.hash(); len(childhash) < 32 {
				nodes[i] = rawNode(childhash)
			}else{
				nodes[i]= hashNode(childhash)
			}
		}
		nodes[16] = nilValueNode
		h = newHasher(false)
		defer returnHasherToPool(h)
		h.tmp.Reset()
		if err := rlp.Encode(&h.tmp, nodes); err != nil {
			panic(err)
		}
```